File: ParameterReader.cpp
#include "ParameterReader.h"
#include "Constants.h"

#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <windows.h>

std::vector<InputParam> ParameterReader::read(std::string &fileName) {
    const char delimiter = PARAMETER_DELIMITER;

    auto pList = std::vector<InputParam>();

    TCHAR buffer[MAX_PATH] = {0};
    GetModuleFileName(nullptr, buffer, MAX_PATH);
    std::string path = std::string(buffer) + R"(\..\SlimeSim\)" + fileName;

    std::ifstream file;
    file.open(path);

    if (file.is_open()) {
        std::string str;
        while (std::getline(file, str)) {
            std::vector<std::string> out;
            std::stringstream ss(str);
            std::string s;
            while (std::getline(ss, s, delimiter)) {
                out.push_back(s);
            }

            if (out.size() >= 2) {
                InputParam param = {
                        .name = out[0]
                };
                try {
                    double d = std::stod(out[1]);
                    param.valueNumber = d;
                } catch (std::invalid_argument &e) {
                    if (out[1] == "true") {
                        param.valueBool = true;
                    } else if (out[1] == "false") {
                        param.valueBool = false;
                    } else {
                        param.valueString = out[1];
                    }
                }
                pList.push_back(param);
            }
        }
    } else {
        std::cout << "FILE FAILED TO OPEN: [" << path << "]" << std::endl;
    }

    file.close();

    return pList;
}

File: ParameterReader.py
from InputParam import InputParam
import Constants


class ParameterReader:
    @staticmethod
    def read(file_name: str) -> list[InputParam]:
        delimiter = Constants.PARAMETER_DELIMITER

        p_list = []
        path = "./{fileName}".format(fileName=file_name)

        try:
            file = open(path)
            while True:
                line = file.readline()

                if line == "":
                    break

                out = line.split(delimiter)

                if len(out) >= 2:
                    param = InputParam()
                    param.name = out[0]
                    try:
                        float_value = float(out[1])
                        param.value_number = float_value
                    except ValueError:
                        bool_value = True if out[1].lower() in Constants.PARAMETER_TRUE_VALUES \
                            else False if out[1].lower() in Constants.PARAMETER_FALSE_VALUES \
                            else None
                        if bool_value is not None:
                            param.value_bool = bool_value
                        else:
                            param.value_string = out[1]

                    p_list.append(param)

            file.close()
        except FileNotFoundError:
            print("FILE FAILED TO OPEN: [{}]\n".format(path))

        return p_list
